### ISSUE TITLE
Strange weight matrix Q

### ISSUE THREAD

**Issue by AlexGidiotis** [Ian09/CF-NADE#10]
I was looking through the NADE code and found a strange weight matrix Q inside the TensorLinear class.

```
output_ = T.tensordot(input_, W, axes=[[1, 2], [0, 1]]) + b
input_mask = T.sum(input_, axis = 2)
output_masked = T.dot(input_mask, Q)
output = output_ + output_masked
 ```

Can anyone explain what this does cause I don't remember reading anything like this in the paper.

**Comment by Ian09** [comment]
Hi Alex,

Thanks for the question. In the first version of the model, I used a Q to model the inputmask. However, I found Q is redundant and can be embedded into W, since we use the cumsum for the input (sharing parameters between different ratings). Hence I removed Q in the model and also the code. You can just ignore the Q of the old version of code. If not, it stills works well since Q plays the same role as the first part of W (i.e. 2x times the learning rate).

**Comment by AlexGidiotis** [comment]
Hi Yin,

Thanks for the quick answer. 

> I found Q is redundant and can be embedded into W, since we use the cumsum for the input (sharing parameters between different ratings).

I assume you are referring to this code:
```
input_ratings_cum = T.extra_ops.cumsum(input_ratings[:, :, ::-1], axis=2)[:, :, ::-1]

y = NADE_CF_model.apply(input_ratings_cum)
```
Is there anything else required for the parameter sharing?

> You can just ignore the Q of the old version of code.

I think it creates some extra trainable parameters that might cause overfitting.

> If not, it stills works well since Q plays the same role as the first part of W (i.e. 2x times the learning rate).

What do you mean with "first part of W" ?

I am trying to implement CF-NADE with Keras and Tensorflow so feel free to have a look at my repo.
[keras-CF-NADE](https://github.com/AlexGidiotis/keras-CF-NADE)

**Comment by Ian09** [comment]
Hi Alex,

The meaning of the "first part of W": suppose W is of shape (N_Ratings, N_Items, Factors), the first part of W is W[0,:,:]. since Q plays the same role of W[0,:,:].

Sorry that the code is not the latest version. please remove the Q variable for the correct implementation.

Thanks!

**Comment by AlexGidiotis** [comment]
Thanks a lot. Did you mean W is of shape (N_Ratings,N_Users,Factors)?

I assume that I should remove these lines:
```
input_mask = T.sum(input_, axis = 2)
output_masked = T.dot(input_mask, Q)
output = output_ + output_masked
```

**Comment by Ian09** [comment]
Yes, you are right! ;-)
